src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp

```diff
@@ -1727,18 +1727,30 @@ size_t ReplicatedMergeTreeQueue::countMutations() const
     return mutations_by_znode.size();
 }
 
-
 size_t ReplicatedMergeTreeQueue::countFinishedMutations() const
 {
     std::lock_guard lock(state_mutex);
 
     size_t count = 0;
-    for (const auto & pair : mutations_by_znode)
+    for (const auto & [_, status] : mutations_by_znode)
     {
-        const auto & mutation = pair.second;
-        if (!mutation.is_done)
+        if (!status.is_done)
             break;
+        ++count;
+    }
+
+    return count;
+}
 
+size_t ReplicatedMergeTreeQueue::countUnfinishedMutations() const
+{
+    std::lock_guard lock(state_mutex);
+
+    size_t count = 0;
+    for (const auto & [_, status] : mutations_by_znode | std::views::reverse)
```

Comment:
I believe we count the mutations that are active (not queued), right ?
so that if we don't have too active, but queued, they are not counted as unfinished. I think it is implemented like this, just wanted to clarify.

Reply:
This counts all unfinished mutations: currently executing and queued.

Reply:
oh then if we have no currently running mutations, and lot of queued mutations, then adding delay will not be helpful. Can we count only the currently executing ones?

Reply:
I think it's better to count all mutations.

We cannot easily determine whether mutation is executing or not. We can say that mutation is not executing if there are no currently executing entries in pool for that mutation or if there are no mutated parts to version of this mutation. In the first case a wait can help because some work can be done in background and some mutation will start to execute. In the second case some parts could be mutated and then something can go wrong and mutations may be actually stuck. 

It's intended to use wait only as a soft threshold with hard threshold after which an exception will be thrown. So, if something is completely broken some time will be spent on waiting but mutations will quickly reach the hard limit. We have almost the same logic for inserts now.

In general waiting for mutations is useless now but it may be useful for lightweight mutations in the future.